File: frontend/src/lib/monitoreo.ts
import { api } from "@/lib/api";

export interface MetricaResumen {
  ts: string;
  ancho_banda_mbps: number;
  latencia_ms: number;
  carga_pct: number;
  clientes_conectados: number;
  estado_semaforo: "good" | "high" | "saturated";
}

export interface AccessPointEstado {
  id: string;
  codigo: string;
  nombre: string;
  edificio_id: string | null;
  edificio_codigo: string | null;
  piso_id: string | null;
  latitud: number | null;
  longitud: number | null;
  activo: boolean;
  banda: string;
  ultima_metrica: MetricaResumen | null;
}

export interface MonitoreoResponse {
  total: number;
  activos: number;
  con_alertas: number;
  access_points: AccessPointEstado[];
}

export interface AlertaOut {
  id: string;
  access_point_id: string;
  ap_codigo: string;
  ap_nombre: string;
  edificio_codigo: string | null;
  tipo: string;
  estado: string;
  umbral_violado: number;
  valor_observado: number;
  detectada_at: string;
  atendida_at: string | null;
  comentario_resolucion: string | null;
}

export const monitoreoApi = {
  async estado(edificio_id?: string): Promise<MonitoreoResponse> {
    const r = await api.get<MonitoreoResponse>("/monitoreo/estado", {
      params: edificio_id ? { edificio_id } : undefined,
    });
    return r.data;
  },
};

export const alertasApi = {
  async listar(solo_activas = true): Promise<AlertaOut[]> {
    const r = await api.get<AlertaOut[]>("/alertas", {
      params: { solo_activas },
    });
    return r.data;
  },

  async atender(
    id: string,
    comentario?: string,
  ): Promise<AlertaOut> {
    const r = await api.patch<AlertaOut>(`/alertas/${id}/atender`, {
      comentario_resolucion: comentario ?? null,
    });
    return r.data;
  },
};

File: backend/app/schemas/monitoreo.py
"""Schemas Pydantic para monitoreo en tiempo real (RF002)."""

from __future__ import annotations

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict

from app.models.access_point import BandaFrecuencia


class MetricaResumen(BaseModel):
    model_config = ConfigDict(use_enum_values=True)

    ts: datetime
    ancho_banda_mbps: float
    latencia_ms: float
    carga_pct: float
    clientes_conectados: int
    estado_semaforo: str  # "good" | "high" | "saturated"


class AccessPointEstado(BaseModel):
    model_config = ConfigDict(use_enum_values=True)

    id: UUID
    codigo: str
    nombre: str
    edificio_id: UUID | None
    edificio_codigo: str | None
    piso_id: UUID | None
    latitud: float | None
    longitud: float | None
    activo: bool
    banda: BandaFrecuencia
    ultima_metrica: MetricaResumen | None


class MonitoreoResponse(BaseModel):
    total: int
    activos: int
    con_alertas: int
    access_points: list[AccessPointEstado]


class MetricaIn(BaseModel):
    access_point_id: UUID
    ancho_banda_mbps: float
    latencia_ms: float
    carga_pct: float
    paquetes_perdidos: int = 0
    clientes_conectados: int = 0
